picard/log.py

```diff
@@ -17,113 +17,206 @@
 from PyQt5 import QtCore
-from picard.util import thread
 
 
-LOG_INFO = 1
-LOG_WARNING = 2
-LOG_ERROR = 4
-LOG_DEBUG = 8
+_MAX_TAIL_LEN = 10**6
 
+VERBOSITY_DEFAULT = logging.WARNING
 
-class Logger(object):
 
-    def __init__(self, maxlen=0):
-        self._receivers = []
-        self.maxlen = maxlen
-        self.reset()
+def debug_mode(enabled):
+    if enabled:
+        main_logger.setLevel(logging.DEBUG)
+    else:
+        main_logger.setLevel(VERBOSITY_DEFAULT)
+
+
+_feat = namedtuple('_feat', ['name', 'prefix', 'fgcolor'])
+
+levels_features = OrderedDict([
+    (logging.ERROR,   _feat('Error',   'E', 'red')),
+    (logging.WARNING, _feat('Warning', 'W', 'darkorange')),
+    (logging.INFO,    _feat('Info',    'I', 'black')),
+    (logging.DEBUG,   _feat('Debug',   'D', 'purple')),
+])
+
+
+# COMMON CLASSES
+
+
+TailLogTuple = namedtuple(
+    'TailLogTuple', ['pos', 'message', 'level'])
+
+
+def _DummyFn(*args, **kwargs):
+    """Placeholder function.
+
+    Raises:
+        NotImplementedError
+    """
+    _, _ = args, kwargs
+    raise NotImplementedError()
 
-    def reset(self):
-        if self.maxlen > 0:
-            self.entries = deque(maxlen=self.maxlen)
-        else:
-            self.entries = []
 
-    def register_receiver(self, receiver):
-        self._receivers.append(receiver)
+# _srcfile is used when walking the stack to check when we've got the first
+# caller stack frame, by skipping frames whose filename is that of this
+# module's source. It therefore should contain the filename of this module's
+# source file.
+_srcfile = os.path.normcase(_DummyFn.__code__.co_filename)
+if hasattr(sys, '_getframe'):
+    def currentframe():
+        return sys._getframe(3)
+else:  # pragma: no cover
+    def currentframe():
+        """Return the frame object for the caller's stack frame."""
+        try:
+            raise Exception
+        except Exception:
+            return sys.exc_info()[2].tb_frame.f_back
 
-    def unregister_receiver(self, receiver):
-        self._receivers.remove(receiver)
 
-    def message(self, level, message, *args):
-        if not self.log_level(level):
-            return
-        if not isinstance(message, str):
-            message = repr(message)
-        if args:
-            message = message % args
-        time = QtCore.QTime.currentTime()
-        message = "%s" % (message,)
-        self.entries.append((level, time, message))
-        for func in self._receivers:
-            try:
-                thread.to_main(func, level, time, message)
-            except:
-                import traceback
-                traceback.print_exc()
+class OurLogger(logging.getLoggerClass()):
 
-    def log_level(self, level):
-        return True
 
+    # copied from https://github.com/python/cpython/blob/3.5/Lib/logging/__init__.py#L1353-L1381
+    # see https://stackoverflow.com/questions/4957858/how-to-write-own-logging-methods-for-own-logging-levels
+    def findCaller(self, stack_info=False):
```

Comment:
Maybe I'm missing something, but why do we need our own copy of this method? Replacing this whole class with

```python
class OurLogger(logging.getLoggerClass()):
    pass
```

appears to result in exactly the same log messages.

Also, technically, there's a license declaration at the top of `logging.py` in cpython that this copy here appears to violate (but I am not a lawyer).

Reply:
We need it because logging module is wrapped (log.* methods) and therefore its findCaller method is returning log method file/lineno etc.. instead of the method executing log.*().


About the license thing i think you're right, it should be copied as well, just above this code (or?).

Reply:
https://github.com/metabrainz/picard/pull/839 removes the useless code ;)